docs: add info about breaking changes to contributing guide (#854)
asyncapi/spec: CONTRIBUTING.md

  The AsyncAPI spec, despite describing technical behavior, is intended to be
  read by people. Use natural tone and include motivation and examples.

## Breaking Change vs Non Breaking Change

> The specification is not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if all suddenly it is not provided. 

Reviewer: ```suggestion
Specifications are not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if all suddenly it is not provided. 
```

_Non-breaking changes:_

> - Adding a property

Reviewer 2: should this be adding an optional property?

Reviewer: Yes, definitely. If you add a required property, it's a breaking change.

Reviewer 3: @dalelane makes sense, can you make a code change suggestion for @Dule-martins please?

Author: @derberg I just committed all suggestions made here.

> - Adding a property

Reviewer 2: ```suggestion
- Adding a new optional property
```

> _Breaking changes:_

Reviewer 2: ```suggestion
_Breaking changes:_
- Adding a new required property
```

- Making a property required
- Making a property optional
- Removing a property

> - Changing the type of a property in any way

Reviewer 4: IMHO this sentence is not precise. I'm suggesting following one:

```suggestion
- Changing the type of a property in backward compatible way
```
**Justification:**: There are multiple ways the field type can change. AsyncAPI 2.x.y is using **UNION** structural typing, thus there are two situations that are similar, but in context of backward compatibility, different.

**Case 1 - backward compatible**

Before Type: `Server Variable Object`
After Type: `Server Variable Object | Reference Object`

**Case 2 - backward incompatible**

Before Type: `Server Variable Object | Reference Object`
After Type: `Server Variable Object`

There is also another case which is variant of `Case 1` and backward compatible as long as `Server Variable Object` and `Difference Object` have the same shape (in context of structural typing). This case can be simplified by calling it a **type renaming**.

**Case 1 (variant 1) - backward compatible**

Before Type: `Server Variable Object`
After Type: `Different Object`

Reviewer: Then your suggestion should say "backward incompatible" way, right? 🤔

Reviewer 4: @fmvilas right ;] Thanks, updated the suggestion.

> - Changing the type of a property in any way

Reviewer: Is it intended to be an exhaustive list or just an example?

Author: from the suggestion made I think it is a list, I can't be sure of how exhaustive it is nor if it is an example.
Maybe @derberg could explain that.

Reviewer 3: it is intended as a result of https://github.com/asyncapi/spec/issues/688

it should be explicit

  The AsyncAPI spec, despite describing technical behavior, is intended to be
  read by people. Use natural tone and include motivation and examples.

## Breaking Change vs Non Breaking Change

> Specifications are not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if all suddenly it is not provided. 

Reviewer 2: ```suggestion
Specifications are not much different than software. Some changes provided in the spec can cause breaking changes for tools that support it. For example, if one of the properties that was `required` becomes `optional`, it is considered a breaking change because some tools might depend on that property and fail if it is not provided. 
```

suggested improvement for readability - I don't think it changes the intended meaning

Author: Yes
